crypto/pkcs7/pkcs7.c

```diff
@@ -191,3 +194,384 @@ int pkcs7_add_signed_data(CBB *out,
+    case NID_pkcs7_signed:
+        PKCS7_free(p7->d.sign->contents);
+        p7->d.sign->contents = p7_data;
+        break;
+    case NID_pkcs7_digest:
+        PKCS7_free(p7->d.digest->contents);
+        p7->d.digest->contents = p7_data;
+        break;
+    case NID_pkcs7_data:
+    case NID_pkcs7_enveloped:
+    case NID_pkcs7_signedAndEnveloped:
+    case NID_pkcs7_encrypted:
+    default:
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_UNSUPPORTED_CONTENT_TYPE);
+        return 0;
+    }
+    return 1;
+}
+
+int PKCS7_content_new(PKCS7 *p7, int type)
+{
+    PKCS7 *ret = PKCS7_new();
+    if (ret == NULL) {
+        goto err;
+    }
+    if (!PKCS7_set_type(ret, type)) {
+        goto err;
+    }
+    if (!PKCS7_set_content(p7, ret)) {
+        goto err;
+    }
+    return 1;
+ err:
+    PKCS7_free(ret);
+    return 0;
+}
+
+int PKCS7_add_recipient_info(PKCS7 *p7, PKCS7_RECIP_INFO *ri) {
+    STACK_OF(PKCS7_RECIP_INFO) *sk;
+    switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_signedAndEnveloped:
+        sk = p7->d.signed_and_enveloped->recipientinfo;
+        break;
+    case NID_pkcs7_enveloped:
+        sk = p7->d.enveloped->recipientinfo;
+        break;
+    default:
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+        return 0;
+    }
+
+    if (!sk_PKCS7_RECIP_INFO_push(sk, ri)) {
+        return 0;
+    }
+    return 1;
+}
+
+int PKCS7_add_signer(PKCS7 *p7, PKCS7_SIGNER_INFO *p7i) {
+    ASN1_OBJECT *obj;
+    X509_ALGOR *alg;
+    STACK_OF(PKCS7_SIGNER_INFO) *signer_sk;
+    STACK_OF(X509_ALGOR) *md_sk;
+
+    switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_signed:
+        signer_sk = p7->d.sign->signer_info;
+        md_sk = p7->d.sign->md_algs;
+        break;
+    case NID_pkcs7_signedAndEnveloped:
+        signer_sk = p7->d.signed_and_enveloped->signer_info;
+        md_sk = p7->d.signed_and_enveloped->md_algs;
+        break;
+    default:
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+        return 0;
+    }
+
+
+    obj = p7i->digest_alg->algorithm;
+    /* If the digest is not currently listed, add it */
+    int alg_found = 0;
+    for (size_t i = 0; i < sk_X509_ALGOR_num(md_sk); i++) {
+        alg = sk_X509_ALGOR_value(md_sk, i);
+        if (OBJ_cmp(obj, alg->algorithm) == 0) {
+            alg_found = 1;
+            break;
+        }
+    }
+    if (!alg_found) {
+        if ((alg = X509_ALGOR_new()) == NULL
+            || (alg->parameter = ASN1_TYPE_new()) == NULL) {
+            X509_ALGOR_free(alg);
+            OPENSSL_PUT_ERROR(PKCS7, ERR_R_ASN1_LIB);
+            return 0;
+        }
+        /*
+         * If there is a constant copy of the ASN1 OBJECT in libcrypto, then
+         * use that.  Otherwise, use a dynamically duplicated copy
+         */
+        int nid = OBJ_obj2nid(obj);
+        if (nid != NID_undef) {
+            alg->algorithm = OBJ_nid2obj(nid);
+        } else {
+            alg->algorithm = OBJ_dup(obj);
+        }
```

Comment:
Although there's no `const` in its signature, it looks like `OBJ_nid2obj` always returns a builtin or static value that should not be freed.(?)

How does `X509_ALGOR_free` know whether or not to free `alg->algorithm`?

Reply:
the `X509_ALGOR` sequence is [defined](https://github.com/aws/aws-lc/blob/3571a86680b191986b6657c4a16383386a091ae3/crypto/x509/x_algor.c#L67) as:

```
ASN1_SEQUENCE(X509_ALGOR) = {
    ASN1_SIMPLE(X509_ALGOR, algorithm, ASN1_OBJECT),
    ASN1_OPT(X509_ALGOR, parameter, ASN1_ANY),
} ASN1_SEQUENCE_END(X509_ALGOR)
```

with the `algorithm` member of type `ASN1_OBJECT`, which tells the macro system to use `ASN1_OBJECT_free` when implementing `X509_ALGOR_free`. `ASN1_OBJECT_free` is [smart enough](https://github.com/aws/aws-lc/blob/3571a86680b191986b6657c4a16383386a091ae3/crypto/asn1/a_object.c#L199-L209) to check `asn1_obj_st`'s [flags](https://github.com/aws/aws-lc/blob/3571a86680b191986b6657c4a16383386a091ae3/crypto/asn1/internal.h#L117) for whether it was dynamically allocated\*, and no-op if not.

\* as you mention [`OBJ_nid2obj` ](https://github.com/aws/aws-lc/blob/3571a86680b191986b6657c4a16383386a091ae3/crypto/obj/obj.c#L354) consults [`kObjects`](https://github.com/aws/aws-lc/blob/main/crypto/obj/obj_dat.h#L7235) to look up a static object to return. all members of `kObjects` have flag (last struct member) [of 0](https://github.com/aws/aws-lc/blob/3571a86680b191986b6657c4a16383386a091ae3/crypto/asn1/internal.h#L117), so none of the [`ASN1_OBJECT_FLAG_DYNAMIC_*` flags](https://github.com/aws/aws-lc/blob/3571a86680b191986b6657c4a16383386a091ae3/crypto/asn1/internal.h#L100) get set.